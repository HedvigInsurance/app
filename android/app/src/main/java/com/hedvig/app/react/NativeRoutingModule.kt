package com.hedvig.app.react

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.content.LocalBroadcastManager
import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.rx2.Rx2Apollo
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.firestore.FirebaseFirestore
import com.hedvig.android.owldroid.graphql.MemberIdQuery
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import timber.log.Timber
import java.util.Date
import java.util.HashMap

class NativeRoutingModule constructor(
    reactContext: ReactApplicationContext,
    private val apolloClient: ApolloClient
) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener {

    private val profileBroadcastReceiver = ProfileBroadcastReceiver()
    private val onBoardingBroadcastReceiver = OnBoardingBroadcastReceiver()

    private val localBroadcastManager = LocalBroadcastManager.getInstance(reactContext)

    private val deviceEventEmitter: DeviceEventManagerModule.RCTDeviceEventEmitter
        get() = reactApplicationContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)

    private val compositeDisposable = CompositeDisposable()

    init {
        reactContext.addLifecycleEventListener(this)
    }

    override fun getName() = "NativeRouting"

    override fun onHostResume() {
        localBroadcastManager.registerReceiver(profileBroadcastReceiver, IntentFilter("profileNavigation"))
        localBroadcastManager.registerReceiver(onBoardingBroadcastReceiver, IntentFilter(ON_BOARDING_INTENT_FILER))
    }

    override fun onHostPause() {
        localBroadcastManager.unregisterReceiver(profileBroadcastReceiver)
        localBroadcastManager.unregisterReceiver(onBoardingBroadcastReceiver)
    }

    override fun onHostDestroy() {
    }

    @ReactMethod
    fun userDidSign() {
        val referralsStorage = currentActivity!!.getSharedPreferences("referrals", Context.MODE_PRIVATE)
        val referee = referralsStorage
            .getString("referee", null)
        val incentive = referralsStorage
            .getString("incentive", null)

        if (referee == null || incentive == null) {
            return
        }

        val memberIdQuery = MemberIdQuery()
        compositeDisposable +=
            Rx2Apollo.from(apolloClient.query(memberIdQuery)).subscribe({ response ->
                val data = response.data()
                if (data == null) {
                    Timber.e("Data was null when trying to accredit referral")
                    return@subscribe
                }
                val member = data.member()
                val memberId = member.id()
                if (memberId == null) {
                    Timber.e("memberId was null when trying to accredit referral")
                    return@subscribe
                }

                val b = Bundle()
                b.putString("invitedByMemberId", referee)
                b.putString("memberId", memberId)
                b.putString("incentive", incentive)

                FirebaseAnalytics
                    .getInstance(reactApplicationContext)
                    .logEvent("referrals_sign", b)

                val referralParameters = HashMap<String, String>()
                referralParameters["invitedByMemberId"] = referee
                referralParameters["memberId"] = memberId
                referralParameters["incentive"] = incentive
                referralParameters["timestamp"] = java.lang.Long.toString(Date().time)

                FirebaseFirestore
                    .getInstance()
                    .collection("referrals")
                    .add(referralParameters)
                    .addOnSuccessListener { Timber.i("Successfully saved referral!") }
                    .addOnFailureListener { error -> Timber.e(error, "Failed to save referral") }

            }, { error -> Timber.e(error, "Failed to load memberId for referral") })
    }

    private fun logoutAndRestartApplication() {
        deviceEventEmitter.emit("NativeRoutingLogoutAndRestartApplication", null)
    }

    private fun restartChatOnBoarding() {
        deviceEventEmitter.emit("NativeRoutingRestartChatOnBoarding", null)
    }

    private inner class ProfileBroadcastReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.getStringExtra(NAVIGATE_ROUTING_EXTRA_NAME_ACTION)) {
                NAVIGATE_ROUTING_EXTRA_VALUE_LOGOUT_AND_RESTART_APPLICATION -> {
                    logoutAndRestartApplication()
                }
            }
        }
    }

    private inner class OnBoardingBroadcastReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.getStringExtra(NAVIGATE_ROUTING_EXTRA_NAME_ACTION)) {
                NAVIGATE_ROUTING_EXTRA_VALUE_LOGOUT_AND_RESTART_APPLICATION -> logoutAndRestartApplication()
                NAVIGATE_ROUTING_EXTRA_VALUE_RESTART_CHAT_ON_BOARDING -> restartChatOnBoarding()
            }
        }
    }

    companion object {

        const val ON_BOARDING_INTENT_FILER = "on_boarding_intent_filter"

        const val NAVIGATE_ROUTING_EXTRA_NAME_ACTION = "action"

        const val NAVIGATE_ROUTING_EXTRA_VALUE_LOGOUT_AND_RESTART_APPLICATION = "logoutAndRestartApplication"
        const val NAVIGATE_ROUTING_EXTRA_VALUE_RESTART_CHAT_ON_BOARDING = "restartChatOnBoarding"
    }
}
