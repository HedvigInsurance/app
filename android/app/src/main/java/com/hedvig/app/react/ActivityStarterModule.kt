package com.hedvig.app.react

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.app.FragmentActivity
import android.support.v4.app.FragmentManager
import android.support.v4.content.LocalBroadcastManager
import androidx.navigation.Navigation
import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.rx2.Rx2Apollo
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.google.firebase.analytics.FirebaseAnalytics
import com.hedvig.android.owldroid.graphql.InsuranceStatusQuery
import com.hedvig.android.owldroid.type.InsuranceStatus
import com.hedvig.app.LoggedInActivity
import com.hedvig.app.R
import com.hedvig.app.feature.chat.UploadBottomSheet
import com.hedvig.app.feature.dashboard.ui.PerilBottomSheet
import com.hedvig.app.feature.dashboard.ui.PerilIcon
import com.hedvig.app.feature.offer.OfferActivity
import com.hedvig.app.feature.offer.OfferChatOverlayFragment
import com.hedvig.app.feature.referrals.PromotionCodeBottomSheet
import com.hedvig.app.util.extensions.proxyNavigate
import com.hedvig.app.util.extensions.setIsLoggedIn
import com.hedvig.app.util.extensions.triggerRestartActivity
import com.hedvig.app.util.react.AsyncStorageNative
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.rxkotlin.plusAssign
import timber.log.Timber

class ActivityStarterModule(
    reactContext: ReactApplicationContext,
    private val apolloClient: ApolloClient,
    private val asyncStorageNative: AsyncStorageNative
) : ReactContextBaseJavaModule(reactContext), LifecycleEventListener {

    private val disposables = CompositeDisposable()

    private val fragmentManager: FragmentManager
        get() = (reactApplicationContext.currentActivity as FragmentActivity).supportFragmentManager

    private val firebaseAnalytics by lazy { FirebaseAnalytics.getInstance(reactApplicationContext) }

    private val localBroadcastManager = LocalBroadcastManager.getInstance(reactContext)

    private val fileUploadBroadcastReceiver = FileUploadBroadcastReceiver()

    private var fileUploadCallback: Promise? = null

    init {
        localBroadcastManager.registerReceiver(fileUploadBroadcastReceiver, IntentFilter(FILE_UPLOAD_INTENT))
    }

    override fun getName() = "ActivityStarter"

    override fun onHostResume() {
    }

    override fun onHostPause() {
    }

    override fun onHostDestroy() {
        localBroadcastManager.unregisterReceiver(fileUploadBroadcastReceiver)
        disposables.clear()
    }

    @ReactMethod
    fun navigateToOfferFromChat() {
        val activity = reactApplicationContext.currentActivity
        if (activity != null) {
            asyncStorageNative.setKey("@hedvig:isViewingOffer", "true")
            currentActivity?.let {
                it.startActivity(Intent(it, OfferActivity::class.java))
            }
        }
    }

    @ReactMethod
    fun navigateToChatFromOffer() {
        currentActivity?.let {
            it.startActivity(Intent(it, OfferActivity::class.java))
        }
    }

    @ReactMethod
    fun navigateToLoggedInFromChat() {
        currentActivity?.let { activity ->
            reactApplicationContext.setIsLoggedIn(true)
            val intent = Intent(activity, LoggedInActivity::class.java)
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
            activity.startActivity(intent)
        }
    }

    @ReactMethod
    fun showOfferChatOverlay() {
        val offerChatOverlayFragment = OfferChatOverlayFragment()
        offerChatOverlayFragment.show(fragmentManager, "OfferChatOverlay")
    }

    @ReactMethod
    fun showPerilOverlay(subject: String, iconId: String, title: String, description: String) {
        PerilBottomSheet.newInstance(subject, PerilIcon.from(iconId), title, description)
            .show(fragmentManager, "perilSheet")
    }

    @ReactMethod
    fun showPromotionOverlay() {
        PromotionCodeBottomSheet.newInstance()
            .show(fragmentManager, "perilSheet")
    }

    @ReactMethod
    fun showFileUploadOverlay(onUpload: Promise) {
        val uploadBottomSheet = UploadBottomSheet()
        uploadBottomSheet.show(fragmentManager, "FileUploadOverlay")
        fileUploadCallback = onUpload
    }

    @ReactMethod
    fun logEvent(eventName: String, map: ReadableMap) =
        firebaseAnalytics.logEvent(eventName, convertMapToBundle(map))

    @ReactMethod
    fun restartApplication() =
        reactApplicationContext.currentActivity?.let {
            it.triggerRestartActivity(it::class.java)
        }

    @ReactMethod
    fun reloadChat() =
        localBroadcastManager.sendBroadcast(Intent(BROADCAST_RELOAD_CHAT))

    @ReactMethod
    fun doIsLoggedInProcedure() {
        disposables += Rx2Apollo
            .from(apolloClient.query(InsuranceStatusQuery()))
            .subscribe({ response ->
                response.data()?.insurance?.status?.let { status ->
                    when (status) {
                        InsuranceStatus.INACTIVE,
                        InsuranceStatus.INACTIVE_WITH_START_DATE,
                        InsuranceStatus.ACTIVE -> navigateToLoggedInFromChat()
                        InsuranceStatus.PENDING,
                        InsuranceStatus.TERMINATED,
                        InsuranceStatus.`$UNKNOWN` -> {
                            // Let the chat take care of this
                        }
                    }
                }
            }, { Timber.e(it) })
    }

    private fun convertMapToBundle(readableMap: ReadableMap): Bundle {
        val bundle = Bundle()
        val iterator = readableMap.keySetIterator()
        while (iterator.hasNextKey()) {
            val key = iterator.nextKey()
            when (readableMap.getType(key)) {
                ReadableType.Boolean -> bundle.putBoolean(key, readableMap.getBoolean(key))
                ReadableType.Number -> bundle.putDouble(key, readableMap.getDouble(key))
                ReadableType.String -> bundle.putString(key, readableMap.getString(key))
                ReadableType.Map -> bundle.putBundle(key, convertMapToBundle(readableMap.getMap(key)))
                else -> {
                    Timber.e("We don't cover null or array")
                }
            }
        }
        return bundle
    }

    private inner class FileUploadBroadcastReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.getStringExtra(FILE_UPLOAD_RESULT)) {
                FILE_UPLOAD_SUCCESS -> {
                    fileUploadCallback?.resolve(intent.getStringExtra(FILE_UPLOAD_KEY))
                        ?: Timber.e("File upload successful but no callback present")
                    fileUploadCallback = null
                }
                FILE_UPLOAD_ERROR -> {
                    fileUploadCallback?.reject("E_NETWORK_ERROR", "failed to upload")
                        ?: Timber.e("File upload failed but no callback present") // TODO improve
                    fileUploadCallback = null
                }
            }
        }
    }

    companion object {
        const val BROADCAST_RELOAD_CHAT = "reloadChat"
        const val REDEEMED_CODE_BROADCAST = "redeemedCode"

        const val BROADCAST_MESSAGE_NAME = "message"

        const val MESSAGE_PROMOTION_CODE_REDEEMED = "promotionCodeRedeemed"

        const val FILE_UPLOAD_INTENT = "file_upload"
        const val FILE_UPLOAD_RESULT = "file_upload_result"
        const val FILE_UPLOAD_SUCCESS = "success"
        const val FILE_UPLOAD_ERROR = "error"
        const val FILE_UPLOAD_KEY = "key"
    }
}
