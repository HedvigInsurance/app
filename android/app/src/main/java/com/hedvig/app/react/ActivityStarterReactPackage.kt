package com.hedvig.app.react

import com.apollographql.apollo.ApolloClient
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.ViewManager
import com.hedvig.app.util.react.AsyncStorageNative
import java.util.ArrayList

class ActivityStarterReactPackage(
    private val apolloClient: ApolloClient,
    private val asyncStorageNative: AsyncStorageNative
) : ReactPackage {

    override fun createNativeModules(
        reactContext: ReactApplicationContext
    ): List<NativeModule> {
        val modules = ArrayList<NativeModule>()
        modules.add(ActivityStarterModule(reactContext, apolloClient, asyncStorageNative))
        return modules
    }

    override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> {
        return emptyList()
    }
}
