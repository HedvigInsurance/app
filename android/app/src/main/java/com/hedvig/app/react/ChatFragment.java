package com.hedvig.app.react;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactRootView;
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
import dagger.android.support.AndroidSupportInjection;
import javax.inject.Inject;

public class ChatFragment extends Fragment implements DefaultHardwareBackBtnHandler {

    @Inject
    ReactInstanceManager reactInstanceManager;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        AndroidSupportInjection.inject(this);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        ReactRootView reactRootView = new ReactRootView(requireContext());
        reactRootView.startReactApplication(reactInstanceManager, "Chat", null);
        return reactRootView;
    }

    @Override
    public void invokeDefaultOnBackPressed() {
        if (getActivity() != null) {
            getActivity().onBackPressed();
        }
    }
}
