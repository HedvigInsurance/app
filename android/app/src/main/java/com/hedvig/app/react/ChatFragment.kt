package com.hedvig.app.react

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.facebook.react.ReactApplication
import com.facebook.react.ReactInstanceManager
import com.facebook.react.ReactNativeHost
import com.facebook.react.ReactRootView
import com.facebook.react.common.LifecycleState
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler

import dagger.android.support.AndroidSupportInjection

class ChatFragment : Fragment(), DefaultHardwareBackBtnHandler {

    private var mReactRootView: ReactRootView? = null

    protected val reactNativeHost: ReactNativeHost
        get() = (activity!!.application as ReactApplication).reactNativeHost

    internal val reactInstanceManager: ReactInstanceManager
        get() = reactNativeHost.reactInstanceManager

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        AndroidSupportInjection.inject(this)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val reactRootView = ReactRootView(requireContext())
        mReactRootView = reactRootView
        reactRootView.startReactApplication(reactInstanceManager, "Chat", arguments)
        return reactRootView
    }

    override fun onDestroy() {
        super.onDestroy()
        if (mReactRootView != null) {
            mReactRootView!!.unmountReactApplication()
            mReactRootView = null
        }
        if (reactInstanceManager.lifecycleState != LifecycleState.RESUMED) {
            reactInstanceManager.onHostDestroy(activity)
            reactNativeHost.clear()
        }
    }

    override fun invokeDefaultOnBackPressed() {
        if (activity != null) {
            activity!!.onBackPressed()
        }
    }
}
