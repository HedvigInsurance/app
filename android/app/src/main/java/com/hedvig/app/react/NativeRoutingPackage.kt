package com.hedvig.app.react

import com.apollographql.apollo.ApolloClient
import com.facebook.react.ReactPackage
import com.facebook.react.bridge.NativeModule
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.ViewManager

class NativeRoutingPackage constructor(private val apolloClient: ApolloClient) : ReactPackage {

    override fun createNativeModules(reactContext: ReactApplicationContext): List<NativeModule> =
        listOf(NativeRoutingModule(reactContext, apolloClient))

    override fun createViewManagers(reactContext: ReactApplicationContext): List<ViewManager<*, *>> = emptyList()
}
