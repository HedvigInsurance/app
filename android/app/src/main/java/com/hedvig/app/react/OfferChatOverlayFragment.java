package com.hedvig.app.react;

import android.app.Dialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.DialogFragment;
import android.support.v4.content.ContextCompat;
import android.support.v4.content.LocalBroadcastManager;
import android.util.DisplayMetrics;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.WindowManager;
import android.widget.FrameLayout;
import android.widget.ImageView;

import androidx.navigation.Navigation;

import com.apollographql.apollo.ApolloClient;
import com.apollographql.apollo.rx2.Rx2Apollo;
import com.facebook.react.ReactApplication;
import com.facebook.react.ReactInstanceManager;
import com.facebook.react.ReactNativeHost;
import com.facebook.react.ReactRootView;
import com.facebook.react.common.LifecycleState;
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler;
import com.hedvig.android.owldroid.graphql.LogoutMutation;
import com.hedvig.app.NativeRoutingModule;
import com.hedvig.app.R;

import org.jetbrains.annotations.NotNull;

import javax.inject.Inject;

import dagger.android.support.AndroidSupportInjection;
import io.reactivex.disposables.CompositeDisposable;
import timber.log.Timber;

public class OfferChatOverlayFragment extends DialogFragment implements DefaultHardwareBackBtnHandler {

    @Inject
    public ApolloClient apolloClient;

    ViewGroup dialogView;
    ReactRootView mReactRootView;

    CompositeDisposable compositeDisposable = new CompositeDisposable();

    LocalBroadcastManager localBroadcastManager;

    @Override
    public int getTheme() {
        return R.style.DialogTheme;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        AndroidSupportInjection.inject(this);
    }

    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog dialog = new Dialog(getActivity());
        dialogView = (ViewGroup) LayoutInflater.from(getActivity()).inflate(R.layout.offer_chat_overlay_dialog, null);
        dialog.getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        dialog.setContentView(dialogView);
        return dialog;
    }

    @Override
    public void onStart() {
        super.onStart();
        localBroadcastManager = LocalBroadcastManager.getInstance(requireContext());

        DisplayMetrics metrics = new DisplayMetrics();
        WindowManager wm = (WindowManager) requireActivity().getSystemService(Context.WINDOW_SERVICE);
        wm.getDefaultDisplay().getMetrics(metrics);
        float heightPixels = metrics.heightPixels;

        Dialog dialog = getDialog();
        if (dialog != null) {
            dialog.getWindow().setGravity(Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL);
            dialog.getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, (int) (heightPixels * 0.9f));
            dialog.getWindow().setBackgroundDrawable(new ColorDrawable(Color.WHITE));
        }
    }

    @Override
    public void setupDialog(@NotNull Dialog dialog, int style) {
        setUpDialogTopBar(dialogView);

        mReactRootView = new ReactRootView(requireContext());
        mReactRootView.startReactApplication(getReactInstanceManager(), "OfferChat", getArguments());
        ((FrameLayout) dialogView.findViewById(R.id.fragmentContainer)).addView(mReactRootView);
    }

    private void setUpDialogTopBar(ViewGroup dialogView) {
        ImageView closeDialogButton = dialogView.findViewById(R.id.closeChatDialog);
        closeDialogButton.setOnClickListener(v -> dismiss());

        ImageView resetButton = dialogView.findViewById(R.id.resetChatButton);
        resetButton.setOnClickListener(v -> {
            final Dialog dialog = new Dialog(requireContext());
            View resetDialogView = LayoutInflater.from(requireContext()).inflate(R.layout.reset_onboarding_dialog, null);
            dialog.getWindow().setBackgroundDrawable(ContextCompat.getDrawable(requireContext(), R.drawable.dialog_background));
            dialog.setContentView(resetDialogView);

            resetDialogView.findViewById(R.id.chatResetDialogNegativeButton).setOnClickListener((view) -> dialog.dismiss());
            resetDialogView.findViewById(R.id.chatResetDialogPositiveButton).setOnClickListener((view) -> {
                dialog.dismiss();
                logout();
            });

            dialog.show();
        });
    }

    private void logout() {
        compositeDisposable.add(Rx2Apollo.from(apolloClient.mutate(new LogoutMutation())).subscribe(data -> {
            loggedout();
        }, e -> {}));
    }

    private void loggedout() {
        Timber.i("logged out >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>");

        Intent resetLogoutIntent = new Intent(NativeRoutingModule.ON_BOARDING_INTENT_FILER);
        resetLogoutIntent.putExtra("action", "logout");
        Timber.i("OnBoardingBroadcastReceiver sendBroadcast");
        localBroadcastManager.sendBroadcast(resetLogoutIntent);
        dismiss();
        Navigation.findNavController(requireActivity(), com.hedvig.app.common.R.id.rootNavigationHost).popBackStack();
    }

    @Override
    public void onPause() {
        super.onPause();
        compositeDisposable.clear();
        Timber.i("Pause!");

    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (mReactRootView != null) {
            mReactRootView.unmountReactApplication();
            mReactRootView = null;
        }
        if (getReactInstanceManager().getLifecycleState() != LifecycleState.RESUMED) {
            getReactInstanceManager().onHostDestroy(getActivity());
            getReactNativeHost().clear();
        }
    }

    protected ReactNativeHost getReactNativeHost() {
        return ((ReactApplication) getActivity().getApplication()).getReactNativeHost();
    }

    ReactInstanceManager getReactInstanceManager() {
        return getReactNativeHost().getReactInstanceManager();
    }

    @Override
    public void invokeDefaultOnBackPressed() {
        if (getActivity() != null) {
            getActivity().onBackPressed();
        }
    }
}
