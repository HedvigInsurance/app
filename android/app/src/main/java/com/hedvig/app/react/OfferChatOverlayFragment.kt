package com.hedvig.app.react

import android.app.Dialog
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.os.Bundle
import android.support.v4.app.DialogFragment
import android.support.v4.content.ContextCompat
import android.support.v4.content.LocalBroadcastManager
import android.util.DisplayMetrics
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import android.view.WindowManager
import android.widget.FrameLayout
import android.widget.ImageView

import androidx.navigation.Navigation

import com.apollographql.apollo.ApolloClient
import com.apollographql.apollo.rx2.Rx2Apollo
import com.facebook.react.ReactApplication
import com.facebook.react.ReactInstanceManager
import com.facebook.react.ReactNativeHost
import com.facebook.react.ReactRootView
import com.facebook.react.common.LifecycleState
import com.facebook.react.modules.core.DefaultHardwareBackBtnHandler
import com.hedvig.android.owldroid.graphql.LogoutMutation
import com.hedvig.android.owldroid.util.extensions.localBroadcastManager
import com.hedvig.app.NativeRoutingModule
import com.hedvig.app.R

import javax.inject.Inject

import dagger.android.support.AndroidSupportInjection
import io.reactivex.disposables.CompositeDisposable
import timber.log.Timber

class OfferChatOverlayFragment : DialogFragment(), DefaultHardwareBackBtnHandler {

    @Inject
    lateinit var apolloClient: ApolloClient

    lateinit var dialogView: ViewGroup
    private var mReactRootView: ReactRootView? = null

    private val compositeDisposable = CompositeDisposable()

    private val reactNativeHost: ReactNativeHost
        get() = (requireActivity().application as ReactApplication).reactNativeHost

    private val reactInstanceManager: ReactInstanceManager
        get() = reactNativeHost.reactInstanceManager

    override fun getTheme(): Int {
        return R.style.DialogTheme
    }

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        AndroidSupportInjection.inject(this)
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = Dialog(activity!!)
        dialogView = LayoutInflater.from(activity).inflate(R.layout.offer_chat_overlay_dialog, null) as ViewGroup
        dialog.window!!.requestFeature(Window.FEATURE_NO_TITLE)
        dialog.setContentView(dialogView)
        return dialog
    }

    override fun onStart() {
        super.onStart()

        val metrics = DisplayMetrics()
        val wm = requireActivity().getSystemService(Context.WINDOW_SERVICE) as WindowManager
        wm.defaultDisplay.getMetrics(metrics)
        val heightPixels = metrics.heightPixels.toFloat()

        val dialog = dialog
        if (dialog != null) {
            dialog.window!!.setGravity(Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL)
            dialog.window!!.setLayout(ViewGroup.LayoutParams.MATCH_PARENT, (heightPixels * 0.9f).toInt())
            dialog.window!!.setBackgroundDrawable(ColorDrawable(Color.WHITE))
        }
    }

    override fun setupDialog(dialog: Dialog, style: Int) {
        setUpDialogTopBar(dialogView)

        mReactRootView = ReactRootView(requireContext())
        mReactRootView!!.startReactApplication(reactInstanceManager, "OfferChat", arguments)
        (dialogView.findViewById<View>(R.id.fragmentContainer) as FrameLayout).addView(mReactRootView)
    }

    private fun setUpDialogTopBar(dialogView: ViewGroup) {
        val closeDialogButton = dialogView.findViewById<ImageView>(R.id.closeChatDialog)
        closeDialogButton.setOnClickListener { v -> dismiss() }

        val resetButton = dialogView.findViewById<ImageView>(R.id.resetChatButton)
        resetButton.setOnClickListener {
            val dialog = Dialog(requireContext())
            val resetDialogView = LayoutInflater.from(requireContext()).inflate(R.layout.reset_onboarding_dialog, null)
            dialog.window!!.setBackgroundDrawable(
                ContextCompat.getDrawable(
                    requireContext(),
                    R.drawable.dialog_background
                )
            )
            dialog.setContentView(resetDialogView)

            resetDialogView.findViewById<View>(R.id.chatResetDialogNegativeButton)
                .setOnClickListener { dialog.dismiss() }
            resetDialogView.findViewById<View>(R.id.chatResetDialogPositiveButton).setOnClickListener {
                dialog.dismiss()
                logout()
            }

            dialog.show()
        }
    }

    private fun logout() {
        compositeDisposable.add(
            Rx2Apollo.from(apolloClient.mutate(LogoutMutation())).subscribe(
                { loggedout() },
                { Timber.e(it) })
        )
    }

    private fun loggedout() {
        Timber.i("logged out >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>")

        val resetLogoutIntent = Intent(NativeRoutingModule.ON_BOARDING_INTENT_FILER)
        resetLogoutIntent.putExtra("action", "logout")
        Timber.i("OnBoardingBroadcastReceiver sendBroadcast")
        localBroadcastManager.sendBroadcast(resetLogoutIntent)
        dismiss()
        Navigation.findNavController(requireActivity(), com.hedvig.app.common.R.id.rootNavigationHost).popBackStack()
    }

    override fun onPause() {
        super.onPause()
        compositeDisposable.clear()
        Timber.i("Pause!")
    }

    override fun onDestroy() {
        super.onDestroy()
        if (mReactRootView != null) {
            mReactRootView!!.unmountReactApplication()
            mReactRootView = null
        }
        if (reactInstanceManager.lifecycleState != LifecycleState.RESUMED) {
            reactInstanceManager.onHostDestroy(activity)
            reactNativeHost.clear()
        }
    }

    override fun invokeDefaultOnBackPressed() {
        if (activity != null) {
            activity!!.onBackPressed()
        }
    }
}
