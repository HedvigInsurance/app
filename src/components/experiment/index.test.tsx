import * as React from 'react';
import { View } from 'react-native';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

import { getGroup, Experiment } from './index';

test('should return something from groups when getting a group', async () => {
  const groups = ['A', 'B'];
  expect(groups).toContain(await getGroup('test', groups));
});

test('should render content', (done) => {
  const element = <View />;

  const groups = ['A', 'B', 'C'];

  const wrapper = mount(
    <Experiment groups={groups} name="mock">
      {(group) => {
        expect(groups).toContain(group);
        return element;
      }}
    </Experiment>,
  );

  process.nextTick(() => {
    wrapper.update();
    expect(wrapper.containsMatchingElement(element)).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
    done();
  });
});
