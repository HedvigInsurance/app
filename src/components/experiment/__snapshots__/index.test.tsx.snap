// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render content 1`] = `
<Component
  groups={
    Array [
      "A",
      "B",
      "C",
    ]
  }
  name="mock"
>
  <Container
    actions={
      Object {
        "setGroup": [Function],
      }
    }
  >
    <InnerContainer
      actions={
        Object {
          "setGroup": [Function],
        }
      }
      initialState={Object {}}
    >
      <View>
        <react-native-mock />
      </View>
    </InnerContainer>
  </Container>
</Component>
`;
